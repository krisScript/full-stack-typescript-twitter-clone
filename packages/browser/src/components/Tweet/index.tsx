import React, { FC, memo } from 'react';
import { Link as NavLink, useLocation } from 'react-router-dom';
import Avatar from 'components/Avatar';
import TweetType from 'types/Tweet';
import getTime from 'utilities/getTime';
import TweetBar from './TweetBar';
import {
  TweetWrapper,
  UserBar,
  Username,
  Handle,
  Text,
  Reply,
  ContentContainer,
  AvatarContainer,
  Time,
  Link,
  Img,
} from './styled';

interface TweetProps {
  tweet: TweetType;
  deleteTweetHandler: (tweetId: string) => void;
}
export const Tweet: FC<TweetProps> = ({ tweet, deleteTweetHandler }) => {
  const location = useLocation();
  const { user, text, date, image, link, reply, _id } = tweet;
  const { username, handle, avatar } = user;
  const milliseconds = new Date().getTime() - new Date(date).getTime();
  const { hours, days, minutes } = getTime(milliseconds);
  return (
    <TweetWrapper data-testid={_id}>
      <AvatarContainer>
        <NavLink to={`/users/${user._id}`} data-testid="profile-link">
          <Avatar avatarURL={avatar} altText={username} />
        </NavLink>
      </AvatarContainer>
      <UserBar>
        <Username>{username}</Username> <Handle>@{handle}</Handle>{' '}
        <Time>
          {days <= 0 ? '' : `${days}d:`}
          {hours}h{hours <= 0 ? `:${minutes}m:` : ''}
        </Time>
        {reply ? (
          <Reply>
            Replying to
            <NavLink
              to={{
                pathname: `/tweet/${reply._id}`,
                state: { tweet: location },
              }}
              data-testid="reply-tweet-link"
            >
              @{reply.user.handle}
            </NavLink>{' '}
          </Reply>
        ) : (
          ''
        )}
      </UserBar>
      <ContentContainer>
        <NavLink
          to={{
            pathname: `/tweet/${tweet._id}`,
            state: { tweet: location },
          }}
        >
          <Text>{text}</Text>
        </NavLink>

        {link ? (
          <Link href={link} target="_blank" rel="noopener noreferrer">
            {link}
          </Link>
        ) : (
          ''
        )}

        {image ? (
          <NavLink
            to={{
              pathname: `/tweet/${tweet._id}`,
              state: { tweet: location },
            }}
          >
            <Img src={image} alt="" />
          </NavLink>
        ) : (
          ''
        )}
      </ContentContainer>
      <TweetBar deleteTweetHandler={deleteTweetHandler} tweet={tweet} />
    </TweetWrapper>
  );
};
export default memo(Tweet);
