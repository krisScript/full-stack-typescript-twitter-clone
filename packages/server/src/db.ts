import mongoose from "mongoose";

const mongoURI: string = `mongodb+srv://${process.env.MONGO_USER}:${process.env.MONGO_PASSWORD}@cluster0-zmcyw.mongodb.net/${process.env.MONGO_DATABASE}?retryWrites=true`;
const db = async (): Promise<void> => {
  try {
    await mongoose.connect(mongoURI, { useNewUrlParser: true });
  } catch (err) {}
};

export default db;
