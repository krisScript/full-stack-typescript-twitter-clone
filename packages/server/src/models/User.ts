import mongoose, { Schema } from 'mongoose';
import User from '@customTypes/User';

const UserSchema: Schema = new Schema({
  username: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  handle: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  email: { required: true, type: String, minlength: 3, unique: true },
  password: { required: true, type: String, minlength: 12 },
  website: { required: false, type: String, minlength: 3 },
  confirmed: { type: Boolean, default: false },
  profilePhoto: { type: String },
  headerPhoto: { type: String },
  date: {
    type: Date,
    default: Date.now,
  },
  following: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  followers: {
    type: Number,
    default: 0,
  },
  likes: [
    {
      source: {
        type: mongoose.Types.ObjectId,
        required: true,
        refPath: 'likes.ref',
      },
      ref: {
        type: String,
        required: true,
        enum: ['Tweet', 'Reply'],
      },
    },
  ],
  bookmarks: [
    {
      source: {
        type: mongoose.Types.ObjectId,
        required: true,
        refPath: 'bookmarks.ref',
      },
      ref: {
        type: String,
        required: true,
        enum: ['Tweet', 'Reply'],
      },
    },
  ],
});

export default mongoose.model<User>('User', UserSchema);
