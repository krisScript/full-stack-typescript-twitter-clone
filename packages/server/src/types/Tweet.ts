import CommonTweet from '@twtr/common/source/types/Tweet';
import { Document } from 'mongoose';

export default interface Tweet extends CommonTweet, Document {
  user: string;
  retweet: string;
  reply: string;
}
