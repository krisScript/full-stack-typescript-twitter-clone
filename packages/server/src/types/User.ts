import CommonUser from '@twtr/common/source/types/User';
import { Schema, Document } from 'mongoose';

export default interface User extends CommonUser, Document {
  password: string;
  confirmed: boolean;
  following: Schema.Types.ObjectId[];
  likes: Schema.Types.ObjectId[];
  bookmarks: Schema.Types.ObjectId[];
}
