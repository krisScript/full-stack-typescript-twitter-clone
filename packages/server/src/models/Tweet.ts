import { NextFunction } from 'express';
import mongoose, { Schema } from 'mongoose';
import Tweet from '@customTypes/Tweet';

const TweetSchema: Schema = new Schema({
  type: { type: String, required: true, enum: ['text', 'image', 'retweet'] },
  text: { type: String, minlength: 1, maxlength: 100 },
  user: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  retweets: { type: Number, default: 0 },
  likes: { type: Number, default: 0 },
  comments: { type: Number, default: 0 },
  date: {
    type: Date,
    default: Date.now,
  },
});

TweetSchema.pre('find', function(next: NextFunction): void {
  this.populate([
    { path: 'user', select: 'username' },
    { path: 'community', select: 'name description subscribers theme' },
  ]);
  next();
});
export default mongoose.model<Tweet>('Tweet', TweetSchema);
