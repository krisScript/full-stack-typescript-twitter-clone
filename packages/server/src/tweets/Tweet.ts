import mongoose, { Schema, Query } from 'mongoose';
import Tweet from '../types/Tweet';

export const TweetSchema: Schema = new Schema({
  type: {
    type: String,
    required: true,
    enum: ['text', 'link', 'retweet', 'reply'],
  },
  text: { type: String, maxlength: 500 },
  image: { type: String },
  link: { type: String },
  user: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  retweet: {
    type: Schema.Types.ObjectId,
    ref: 'Tweet',
  },
  reply: { type: Schema.Types.ObjectId, ref: 'Tweet' },
  retweets: { type: Number, default: 0 },
  likes: { type: Number, default: 0 },
  replies: { type: Number, default: 0 },
  date: {
    type: Date,
    default: Date.now,
  },
});
//@ts-ignore
async function populateFields(this: Query<any>): Promise<void> {
  this.populate([
    { path: 'user', select: 'username handle avatar' },
    { path: 'reply', select: 'user' },
    { path: 'retweet' },
  ]);
}

TweetSchema.pre('find', populateFields);

TweetSchema.pre('populate', populateFields);

export default mongoose.model('Tweet', TweetSchema);
