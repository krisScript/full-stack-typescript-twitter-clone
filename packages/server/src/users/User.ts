import mongoose, { Schema } from 'mongoose';
import User from '../types/User';
import duplicationErrorHandler from '../middleware/duplicationErrorHandler';
//@ts-ignore
import uniqueValidator from 'mongoose-unique-validator';

const UserSchema: Schema = new Schema({
  username: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  handle: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  email: { required: true, type: String, minlength: 3, unique: true },
  password: { required: true, type: String, minlength: 12 },
  isConfirmed: { type: Boolean, default: true },
  avatar: {
    type: String,
  },
  cover: { type: String },
  date: {
    type: Date,
    default: Date.now,
  },
  following: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  followers: {
    type: Number,
    default: 0,
  },
  likes: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],
  retweets: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],
  replies: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],

  bookmarks: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],
});

UserSchema.index({ handle: 'text' });

UserSchema.plugin(uniqueValidator);

export default mongoose.model<User>('User', UserSchema);
