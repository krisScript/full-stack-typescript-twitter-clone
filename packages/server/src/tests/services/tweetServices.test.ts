import mongoose from 'mongoose';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import {
  createTweet,
  createLinkTweet,
  createImageTweet,
  getTweetById,
} from '@services/tweetServices';
import User from '@models/User';
import Tweet from '@models/Tweet';
import db from 'src/db';

jest.mock('@utilities/sendEmail');
jest.mock('jsonwebtoken');

const mockToken = 'mockToken';
(jwt.sign as jest.Mock).mockReturnValue(mockToken);

describe('tweetServices', (): void => {
  beforeAll(
    async (): Promise<void> => {
      db();
      await Tweet.deleteMany({}).exec();
      await User.deleteMany({}).exec();
    },
  );
  afterEach(
    async (): Promise<void> => {
      await Tweet.deleteMany({}).exec();
      await User.deleteMany({}).exec();
    },
  );
  afterAll(
    async (): Promise<void> => {
      await mongoose.disconnect();
    },
  );
  const text =
    'Lorem ipsum dolor sit amet consectetur adipisicing elit. Similique vel alias, amet corporis modi corrupti.';
  const link = 'fakeUrl';
  const imagePath = 'fakeImagePath';
  const mockUserId = mongoose.Types.ObjectId().toString();
  describe('createTweet', (): void => {
    it(`should create a new tweet`, async (): Promise<void> => {
      expect.assertions(3);
      const { tweetId } = await createTweet(text, mockUserId);
      expect(tweetId).toBeTruthy();
      const tweet = await Tweet.findById(tweetId);
      if (!tweet) {
        return;
      }
      expect(tweet.text).toMatch(text);
      expect(tweet.user.toString()).toMatch(mockUserId);
    });
  });
  it('should throw an error', async (): Promise<void> => {
    expect.assertions(1);
    const invalidUserId = 'invalid';
    await expect(createTweet(text, invalidUserId)).rejects.toThrow();
  });
  describe('createImageTweet', (): void => {
    it(`should create a new tweet`, async (): Promise<void> => {
      expect.assertions(4);
      const hashMock = jest.spyOn(bcrypt, 'hash');
      const { tweetId } = await createImageTweet(text, imagePath, mockUserId);
      expect(tweetId).toBeTruthy();
      const tweet = await Tweet.findById(tweetId);
      if (!tweet) {
        return;
      }
      expect(tweet.text).toMatch(text);
      expect(tweet.image).toMatch(imagePath);
      expect(tweet.user.toString()).toMatch(mockUserId);
      hashMock.mockRestore();
    });
  });
  it('should throw an error when the user id is invalid', async (): Promise<
    void
  > => {
    expect.assertions(2);
    const invalidUserId = 'invalid';
    await expect(
      createImageTweet(text, imagePath, invalidUserId),
    ).rejects.toThrow();
    await expect(
      createImageTweet(text, imagePath, invalidUserId),
    ).rejects.toMatchSnapshot();
  });
  describe('createLinkTweet', (): void => {
    it(`should create a new tweet`, async (): Promise<void> => {
      expect.assertions(4);
      const hashMock = jest.spyOn(bcrypt, 'hash');
      const { tweetId } = await createLinkTweet(text, link, mockUserId);

      expect(tweetId).toBeTruthy();
      const tweet = await Tweet.findById(tweetId);
      if (!tweet) {
        return;
      }
      expect(tweet.text).toMatch(text);
      expect(tweet.link).toMatch(link);
      expect(tweet.user.toString()).toMatch(mockUserId);
      hashMock.mockRestore();
    });
  });
  it('should throw an error when the user id is invalid', async (): Promise<
    void
  > => {
    expect.assertions(1);
    const invalidUserId = 'invalid';
    await expect(
      createLinkTweet(text, imagePath, invalidUserId),
    ).rejects.toThrow();
  });
  describe('getTweetById', (): void => {
    it(`should get a tweet`, async (): Promise<void> => {
      expect.assertions(3);
      const type = 'text';
      const userId = mongoose.Types.ObjectId().toString();
      const newTweet = new Tweet({
        text,
        user: userId,
        type,
      });
      await newTweet.save();
      const tweetId = newTweet._id;
      const { tweet } = await getTweetById(tweetId);
      if (!tweet) {
        return;
      }
      expect(tweet.text).toMatch(text);
      expect(tweet.type).toMatch(type);
      expect(tweet.user.toString()).toMatch(userId.toString());
    });

    it('should throw an error when the tweet is not found', async (): Promise<
      void
    > => {
      expect.assertions(2);
      const tweetId = mongoose.Types.ObjectId().toString();
      await expect(getTweetById(tweetId)).rejects.toThrow();
      await expect(getTweetById(tweetId)).rejects.toMatchSnapshot();
    });
  });
});
