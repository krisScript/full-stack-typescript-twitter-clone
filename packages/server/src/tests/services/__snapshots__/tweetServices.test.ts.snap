// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tweetServices createImageTweet should throw an error when the user id is invalid 1`] = `[ValidationError: Tweet validation failed: user: Cast to ObjectID failed for value "invalid" at path "user"]`;

exports[`tweetServices getTweetById should throw an error when the tweet is not found 1`] = `[Error: Resource not found]`;
